core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -839,15 +797,75 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
             .update(err, state)
             .flatMap(
               decision =>
-                if (decision.cont) clock.sleep(decision.delay) *> loop(decision.state)
-                else orElse(err, decision.finish()).map(Left(_))
+                decision.delay.run.flatMap { delay =>
+                  if (decision.cont) clock.sleep(delay) *> loop(decision.state)
+                  else orElse(err, decision.finish()).map(Left(_))
+                }
             ),
         succ => ZIO.succeedRight(succ)
       )
 
     policy.initial.flatMap(loop)
   }
 
+  /**
+   * Schedule this effect using the given schedule.
+   * Scheduled effects will run only when schedule signal it can run, keeping it dormant while it can't or until it fails.
+   */
+  final def scheduled[R1 <: R, B](schedule: ZSchedule[R1, Option[A], B]): ZIO[R1 with Clock, E, B] =
+    scheduleOrElse[R1, E, B](schedule, (e, _) => ZIO.fail(e))
+
+  /**
+   * Schedule this effect with the specified schedule, until it fails, and then both the
+   * value produced by the schedule together with the last error are passed to
+   * the recovery function.
+   * Scheduled effects will run only when schedule signal it can run, keeping it dormant while it can't.
+   */
+  final def scheduleOrElse[R1 <: R, E2, B](
+    schedule: ZSchedule[R1, Option[A], B],
+    orElse: (E, Option[B]) => ZIO[R1, E2, B]
+  ): ZIO[R1 with Clock, E2, B] =
+    scheduleOrElseEither[R1, B, E2, B](schedule, orElse).map(_.merge)
+
+  /**
+   * Schedule this effect with the specified schedule, until it fails, and then both the
+   * value produced by the schedule together with the last error are passed to
+   * the recovery function.
+   * Scheduled effects will run only when schedule signal it can run, keeping it dormant while it can't.
+   */
+  final def scheduleOrElseEither[R1 <: R, B, E2, C](
+    schedule: ZSchedule[R1, Option[A], B],
+    orElse: (E, Option[B]) => ZIO[R1 with Clock, E2, C]
+  ): ZIO[R1 with Clock, E2, Either[C, B]] = {
```

Comment:
The only (non-semantical) difference between these functions and the `repeat*` functions is the `Option[A]` that is fed into the `Schedule`. The `Option[A]` is currently required because you need to get the "first delay" out of the `Schedule`.

However, I think with a small change to `Schedule`, we can make your new semantic the default, and restore the original type as `A` instead of `Option[A]`.

How?

By having `Schedule`'s `initial` state return a _tuple:

```scala
val initial: ZIO[R, Nothing, (Delay, State)]
```

Now as soon as the initial state is computed, the initial delay ("first delay") is also computed; so schedules now have a chance to delay the first execution of the effect.

With this change, we can remove the `schedule*` methods, and change the `repeat*` methods to have your new semantic: they will first respect an initial delay, rather than running _right away_.

Do you think this could work?

Reply:
Yup, definitiely I think it work. Already started implementation and for now it's very promising. Do you think I need to replicate this behavior to retry as well or not?

Reply:
Yes, I think the correct behavior for retry will be: try right away, if it fails, THEN compute initial state, wait first delay, and THEN feed failure into update and begin/continue loop.

We always have to respect initial delay and always feed all A into schedule without skipping any.